Refactor rules into configurable phases

borkaehw opened the pull request:
### Description
<!-- Mandatory: A crisp one or two line description of your proposed change. -->
This is the first step we would like to take to merge bazelbuild/rules_scala and higherkindness/rules_scala. We are turning the implementation part into pipeline.

Details in this [design doc](https://docs.google.com/document/d/1_HMUICdMFdarzqwJ9Nz3kcQwCPF5yjRdX1umisILn8k/edit?usp=sharing). Feel free to leave comment in the design doc or this PR.

The massive PR is just so people can see the full idea but that we're happy to break it up into multiple PRs.

<!-- Optional:
  A longer explanation of your proposed changes..
  This includes listing any breaking changes, if there are any.
-->

### Motivation
<!-- Mandatory: A summary of why you are making this change. -->
We learned a good lesson from Andy Scott's work of converting higherkindness/rules_scala into configurable phases. This will help us to move custom functionality into bazelbuild/rules_scala.

SrodriguezO commented on test/phase/customizability_test.bzl:
```diff
@@ -0,0 +1,35 @@
+        ),
+    ]
+
+add_phase_customizability_test_singleton = rule(
```
It'd be nice to move all this into a macro so that consumers don't have to define so many things to customize rules. I think we can do this later.

SrodriguezO replied:
Nvm I don't think you can define rules in a macro :man_facepalming:  
I'll see if there's a different way to make this cleaner from a consumer's standpoint

borkaehw commented:
@andyscott Could you also take a look when you get a chance? We would like to have your feedback since you are the original author of phase architecture. :smiley:

ianoc commented:
Is this not really just a different implementation of scala rules looking at the size of the replacement?

borkaehw commented:
@ianoc There are two commits.

The first one is all about moving code around, grouping common code together. For example, We grouped related helper functions that used to be in `rule_impl.bzl` into `phase_compile.bzl`. This way we can make them private functions and it's clear they are not used in other places. We are able to simplify code further more by moving code in `rule_impe.bzl` and `common.bzl` to the file that those functions are only imported from, but it will make this PR bigger. We could do that in other PR.

The second one is about implementing the customizability of phase architecture. It includes some changes to the first commit and tests that could also be an example of how we can add zinc compilation step in the future.

So the answer is no, this is not just a different implementation. It also introduces a more flexible structure for future features.

ianoc commented:
On the face of it i'd be strongly -1 of a large change like this. Its re-learning the world for all existing contributers and effectively re-doing these rules to match the other ones from what I can see. At some point if they follow different designs/philosophies should they not just remain separate?

SrodriguezO commented:
@ianoc The spirit of this PR is to incorporate an architecture that makes the rules much more customizable while increasing maintainability in the process.

>effectively re-doing these rules to match the other ones from what I can see

Rest assured that this is _not_ the goal. We learned some lessons from `higherkindness/rules_scala`--like this phases architecture--which we think will add much value to `bazelbuild/rules_scala` in the long term. 

From the design doc:
```
This will provide two major benefits:
  1) it will be easier for contributors to implement new functionality by creating additional default phases. Phases also give us more clear idea what steps are shared across rules.
  2) consumers will be able to configure the rules to their specific use cases by defining new phases within their workspace without impacting other consumers.
```

Although superficially it may seem like we're trying to simply import `higherkindness/rules_scala`, the real goal is to improve the configurability and maintainability of the rule set.

>Its re-learning the world for all existing contributers

The phases @borkaehw used parallel existing helper methods, so there won't be much to re-learn. The phases architecture itself isn't terribly complicated.

borkaehw commented:
We can keep the helper functions unmoved but still achieve the same goal. It will greatly reduce the size of changes.

Although I believe it's a good opportunity clean up the code base, we can always do it later.

ittaiz commented:
Thanks.
To be honest I'm still not sure what I think of this direction (not saying it's bad at all but very different so I need some time) but I greatly appreciate the effort.
While @ianoc's comment definitely resonate with me I'm trying to get used to the phases architecture and evaluate it on its own merits. 

I'll reply next week with more thoughts and details.

SrodriguezO commented:
Sounds great. Thanks @ittaiz

long-stripe commented:
@borkaehw I agree with the notion of unbundling the moving helper functions from this PR - anything to reduce the size of this diff and make it more reviewable. If it is at all possible, to make it more digestible you could even think about stacking PRs, but not a dealbreaker to me.

I'll try and take an earnest look at (1) the design doc, and (2) this PR in the coming days.

borkaehw commented:
> you could even think about stacking PRs

@long-stripe Do you mean having multiple commits in same PR?

long-stripe commented:
I mean something like https://graysonkoonce.com/stacked-pull-requests-keeping-github-diffs-small/, since rules_scala usually squash merges, so even if you have multiple commits in a PR those get fused into one into master. Not sure how this all plays with squash merges though.

For the intent of reviewing, multiple commits that replay a story would also help, but only if it makes sense.

borkaehw commented:
I have undone roughly 600 lines of code changes, it brings the diff from `+1,806 −1,275` to `+1,232 −630`. Most of the deletion comes from the removal of `scala_binary_common` and `_lib`, as well as simplifying `RULE_impl`s.

I will try to break it up more, but hopefully it's now easier to review than original PR.

borkaehw commented:
@long-stripe I have read through the stacked PRs article. I have reduced the changes and broken down the PR into 4 commits, in a reasonable way I think. Are we still in favor of stacked PRs?

long-stripe commented:
@borkaehw not categorically! I was only suggesting something like it, broken up commits is probably also good too.

ittaiz commented:
@borkaehw is this ready?
I reviewed a third of the PR in it's initial state and then you force pushed (to accommodate requests from others) and I now need to start all over. I want to make sure you don't have anything else cooking.

This is a large and complex PR and I don't want us to fall into the pitfall of having a LGTM because the code is too large to ingest. We need to understand all of the changes and it will take some time (obviously I'll make sure not to merge any other PRs and cause merge conflicts in the meantime).

andyscott commented:
This looks like a reasonable port of both code and design approach that I did for higherkindess/rules_scala. I'm particularly bullish on using phases for organizing the rule implementation. They bring flexilibity and organization to rule implementations. That being said, I'm still not sold on my own approach for phase replacements (for this PR, as well as for the current code in higherkindness/rules_scala). It might be simpler to just let folks assemble their own custom rules and call the phases à la carte.

borkaehw commented:
Pushed a commit to break up `phase_init` to multiple more explicit phases.

All of them are just calling one helper function, we could clean it up by moving helper functions directly to phase files in later PR

borkaehw commented:
Pushed a commit to divide phases into configurable and non-configurable. It will prevent consumers from accidentally swapping the mandatory phase, like `phase_final`.

borkaehw commented:
Force pushed to fix lint.

ittaiz commented:
Hi,
Just to update- I'm working through the PR. It's taking time since I'm trying to understand the patterns, review them and caveats, review potential regressions and some layout/formatting/etc.
It's coming along but will take some more time, don't despair ;)

borkaehw commented:
@ittaiz here is an example to show the beauty of phase architecture. We added a formatting phase without changing any main implementation of the rules.

Hopefully it helps with the review process https://github.com/bazelbuild/rules_scala/pull/874

ittaiz commented on scala/private/phases/phase_scala_provider.bzl:
```diff
@@ -0,0 +1,45 @@
+        compile_jars = compile_jars,
+        deploy_jar = ctx.outputs.deploy_jar,
+        full_jars = p.compile.full_jars,
+        ijar = p.compile.class_jar,  # we aren't using ijar here
```
this looks like a bug, no? Doesn't the [lib](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-a7feab8f41e78c82c1e78f85a9fc48eeL118) impl pass ijars?

borkaehw replied:
It does look like a bug. Fixed.

borkaehw replied:
Fixed.

ittaiz commented on scala/private/phases/phase_scala_provider.bzl:
```diff
@@ -0,0 +1,45 @@
+    return _phase_scala_provider(
+        ctx,
+        p,
+        _args.rjars if hasattr(_args, "rjars") else p.compile.rjars,
```
did we lose this https://github.com/bazelbuild/rules_scala/pull/865/files#diff-d68dd15f790bd2a815ab04d3d0a320d8L843?

borkaehw replied:
That was handled in `phase_compile.bzl`
https://github.com/bazelbuild/rules_scala/pull/865/files#diff-78d7aa1b58ceeb6b93b4a167504e05c4R140

andyscott commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,47 @@
+    for (relation, peer_name, name, function) in adjustments:
+        for idx, (needle, _) in enumerate(phases):
+            if needle == peer_name:
+                if relation in ["-", "before"]:
```
The phase extension mechanism is perhaps a bit overkill. It does allow some neat exploration by wrapping up scalac plugins with providers that augment the phases. This way, for example, the semanticdb plugin could intelligently augment the rules and save the semantic database file as an additional output.

This same functionality can also be achieved by simply building your own rule a la carte using the phases directly.

The TLDR is that I like the phase extension mechanism, but I don't know if it has fully proven its worth. If it gets merged as part of this PR I propose we evaluate keeping/removing it at a later date once we have more data.

ittaiz replied:
My original comments were mainly about the operators (symbols in addition to keywords) but I think my more general concern is that we are making rules_scala API dramatically larger by adding all phases, extension mechanism (both via keywords and via symbols).
I'm concerned about maintenance costs.
Current PR also makes it really easy for someone to customize in places where maybe they shouldn't. I'm trying to think how to make customization possible but less accessible so that we don't have many splinters which mean we can't progress.
I'm not sure exactly what this means but one example can be to not export all the phases everywhere, clearly separate phases (files as well as docs) from "regular" API (maybe this is the way it is now I'm not sure since the PR is huge so I'm a bit lost in the details) and another
example is that I'd consider dropping the symbols (not my biggest concern but an example).
wdyt?

borkaehw replied:
> are all of these operators covered in tests?
are they documented for user extensibility?

Not yet, but I am happy to do so.

> I'd consider dropping the symbols

I am fine with it, but I don't think it matters too much.

> easy for someone to customize in places where maybe they shouldn't

People may easily break things if they don't fully understand how things work, but I don't think it should prevent us from making it really flexible for those who have the knowledge. We still provide the official default rules, and that's the part we should always make sure it's awesome.

Phases extension mechanism also makes maintainers/developers easy to make changes. Check out the example for formatting https://github.com/bazelbuild/rules_scala/pull/874.

ittaiz replied:
@borkaehw I think I'm having a hard time getting my point across.
I'm in favor of the phases and even the extension mechanism seems valuable for exactly the reasons you mention.
What I'm concerned about is making our API surface much bigger.
Maybe there's a way we can signal "danger zone" for most users (who obviously don't need this level of extensions).
Does that make sense?

borkaehw replied:
Sure, I understand your concern

> making our API surface much bigger

IMHO stating it clear in the documents and READMEs might be sufficient to prevent bad experience from general users. I could be wrong.

> we can signal "danger zone" for most users

I am interested in how it will look like. :smile:

ittaiz replied:
One suggestion is to have all extension related under `//scala/advanced-usage`. This is to signal it's not part of the standard d2d usage for most companies.
Also, are the symbols here based on some kind of convention (npm or something) or something we made up?

borkaehw replied:
> //scala/advanced-usage

I like it. I will update the PR.

> are the symbols here based on some kind of convention

I think @andyscott would be a better person for this question.

andyscott commented on scala/private/phases/phases.bzl:
```diff
@@ -0,0 +1,116 @@
+load(
```
It's mostly just for convenience. Regarding the underscore, I'm pretty sure you have to do it this way (or similar) to appease modern Bazel.

ittaiz replied:
Re exporting you're right, I thought they kept a way to say export is intentional but they haven't https://github.com/bazelbuild/bazel/issues/5636
Re convenience- the downside of it is coupling since this file depends on basically everything. Not sure here...

borkaehw replied:
It's mostly for convenience +1.

Currently if we do 
```
load(
    "@io_bazel_rules_scala//scala/private:phases/phases.bzl",
    "extras_phases",
    "phase_binary_compile",
    "phase_binary_final",
)
```
without it we have to do
```
load(
    "@io_bazel_rules_scala//scala/private:phases/api.bzl",
    "extras_phases",
)
load(
    "@io_bazel_rules_scala//scala/private:phases/phase_compile.bzl",
    "phase_binary_compile",
)
load(
    "@io_bazel_rules_scala//scala/private:phases/phase_final.bzl",
    "phase_binary_final",
)
```

borkaehw commented on scala/private/phases/phase_init.bzl:
```diff
@@ -0,0 +1,54 @@
+#
+# PHASE: init
```
It's outdated. It has been split to multiple phases already. Please check the latest changes.

borkaehw commented on test/phase/phase_customizability_test.bzl:
```diff
@@ -0,0 +1,7 @@
+# A dummy test phase to make sure rules are customizable
+#
+def phase_customizability_test(ctx, p):
+    print("customizable phase")
```
It's outdated. It is now testing `ctx`. Please check the latest changes.

We could also test `p` if we make the test more complicated.

borkaehw commented on test/phase/BUILD:
```diff
@@ -0,0 +1,23 @@
+    "customizability_test_scala_library",
+)
+
+add_phase_customizability_test_singleton(
```
It is used in `test/phase/customizability_test.bzl`. `phase_providers` uses it to pass in custom phase.

borkaehw commented on test/shell/test_helper.sh:
```diff
@@ -60,6 +60,25 @@ action_should_fail_with_message() {
   fi
 }
 
+action_should_contain_message() {
```
It's outdated. We no longer need this. Please check the latest changes.

borkaehw commented on scala/providers.bzl:
```diff
@@ -57,3 +57,10 @@ declare_scalac_provider = rule(
 )
+
+ScalaRulePhase = provider(
+    doc = "A Scala compiler plugin",
```
I agree it's not the best description. How about `A custom phase plugin`?

ittaiz replied:
👍

borkaehw replied:
Fixed.

borkaehw commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,51 @@
+                    phases[idx] = (name, function)
+    return phases
+
+def run_phases(ctx, customizable_phases, fixed_phase):
```
Sounds like a good idea.

borkaehw replied:
Fixed.

borkaehw commented on scala/private/phases/phase_collect_jars.bzl:
```diff
@@ -0,0 +1,71 @@
+    "collect_jars_from_common_ctx",
+)
+
+def phase_test_collect_jars(ctx, p):
```
I used the convention like
`scala_test` -> `phase_test_collect_jars`
`scala_repl` -> `phase_repl_collect_jars`

If we make it `phase_scala_test_collect_jars`, should we also make it `phase_scala_repl_collect_jars`? I feel like it will significantly increase the phase name but all rules start with `scala_`. Why do you think it's ambiguous?

Or we should come up with a better naming convention?

ittaiz replied:
scala_test is indeed a problematic one, it should probably have been scalatest_test but it was the first one so it grabbed the ambigous name. I think repl can stay as is but I thought that this phase is general for all tests while it's not so for junit scala tests right? This is what I think we should namespace it. Maybe `phase_scalatest_collect_jars`?

borkaehw replied:
I think these other 4 phases also fall into this category
`phase_test_compile`
`phase_test_final`
`phase_test_runfiles`
`phase_test_write_executable`

Should we also change them from `_test_` to `_scalatest_`?

borkaehw replied:
No worries. Fixed.

borkaehw commented:
@ittaiz thanks for all the feedback. If I haven't responded to certain comments, that means I need more time to think of your questions (not ignoring them).

For some reason, you still commented on old commits? I see you responded to the changes that are no longer part of this PR.

We are happy to write documentations, both in-lines and README, and tests to make the phase architecture more maintainable and robust. While, we would like to know if we are moving forward before making all this effort, in case we end up dropping the entire PR. Hopefully it makes sense.

borkaehw commented on scala/scala.bzl:
```diff
@@ -60,3 +67,11 @@ scala_repl = _scala_repl
 scala_test = _scala_test
 scala_test_suite = _scala_test_suite
+
+make_scala_binary = _make_scala_binary
```
User who wants the vanilla mainstream usage can still access the rules in the same file the same way as before. This is indeed a power user tool, but I don't see why don't we provide it.

borkaehw replied:
I like the suggestion.

borkaehw commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,51 @@
+
+    return current_provider
+
+def extras_phases(extras):
```
We will document it for sure. What would be a better name for you?

borkaehw commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,51 @@
+    ]
+
+    if phase_providers != []:
+        customizable_phases = _adjust_phases(customizable_phases, [p for pp in phase_providers for p in pp.phases])
```
Do you mean make the "for comprehension" part a separate private method?

borkaehw commented on scala/providers.bzl:
```diff
@@ -57,3 +57,10 @@ declare_scalac_provider = rule(
+ScalaRulePhase = provider(
+    doc = "A Scala compiler plugin",
+    fields = {
+        "phases": "the phases to add",
```
More documentations and tests for sure.

Do you think there is a way to improve the double level of customization? Would you prefer we make it simpler?

ittaiz replied:
Re double level see comment in summary of review.
Concretely we need to document the vector `(relation, peer_name, name, function)` here since it's the API.
Also, should the field be called `customizable_phases` or `custom_phases`?

borkaehw replied:
I prefer `custom_phases` to keep it shorter. I will update the PR.

borkaehw commented on scala/private/phases/phase_merge_jars.bzl:
```diff
@@ -0,0 +1,19 @@
+    "merge_jars",
+)
+
+def phase_merge_jars(ctx, p):
```
In general, I want to avoid too much of renaming and refactoring to make this PR easier for review. I agree we should name it after it's responsibility and also we should move the helper functions into these phases (no need for a call to other files).

I like the idea but how about we handle it in another PR as well?

borkaehw commented on scala/private/phases/phase_collect_jars.bzl:
```diff
@@ -0,0 +1,71 @@
+    )
+    return phase_common_collect_jars(ctx, p, args)
+
+def phase_common_collect_jars(ctx, p, _args = struct()):
```
You may have noticed we provide this kind of customization in a bunch of phases. To be more specific, phases that need a bit of customization to different rules will need this struct. This is how we transformed function calls into phases. Though I admit it's not ideal.

Documentation is needed for sure.

I wanted to make the `_args` private but it turned out to be unnecessary.

borkaehw commented on scala/private/phases/phase_collect_jars.bzl:
```diff
@@ -0,0 +1,71 @@
+#
+# DOCUMENT THIS
+#
+load(
```
I am not quite sure what's the concerns here. All of the functions are for customizing arguments for different rules, but eventually they call the same helper function `collect_jars_from_common_ctx`.

> I'd rather have more small files than have files that change for many different reasons.

Could you elaborate more on this?

ittaiz replied:
I mean that there are two approaches I can see to organizing these calls:
1. Group all collect_jars (as well as other phases) in the same file since they call the same underlying mechanism.
2. Separate each collect_jar phase into it's own separate file. This way if the way we collect jars for the repl changes (`phase_repl_collect_jars`) junit code is unrelated (`phase_junit_test_collect_jars`).

I know that even in option 1 the diff will only show changes in the repl method but this leads to bigger files which handle a few things.

Is my intention clearer now?

borkaehw replied:
Thanks for clarification! I have thought of the second approach. The downside is we will have several small files with the same purpose, it may make it harder to compare the difference across rules. `phase_compile.bzl` will be the worse case since there are 7 rules with 7 different implementations. I can imagine if I want to make some common changes or inspect the difference, I will need to modify/open 7 files (not ideal for development).

Making subfolders may help with navigation a bit.

> even in option 1 the diff will only show changes in the repl method

I agree with this so I prefer option 1.

borkaehw commented on scala/private/phases/phase_write_manifest.bzl:
```diff
@@ -0,0 +1,12 @@
+)
+
+def phase_write_manifest(ctx, p):
+    write_manifest(ctx)
```
If the users want to customize writing manifest, they have to write their own phase. Therefore, they can't use the built-in `write_manifest`, not even build on top of it, but they have access to `ctx` and all the results from previous phases via `p`.

Do I answer your question?

borkaehw commented on scala/private/phases/phase_write_executable.bzl:
```diff
@@ -0,0 +1,72 @@
+    return _phase_write_executable(
+        ctx,
+        p,
+        _args.rjars if hasattr(_args, "rjars") else p.compile.rjars,
```
Everything goes after `else` is treated as default, but not all defaults come from `p`, so I don't think overriding `p` is a good. Also, overriding `p` essentially means we might be passing phase specific things to future phases, which should be kept private.

borkaehw commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,51 @@
+    for (name, function) in customizable_phases + [fixed_phase]:
+        new_provider = function(ctx, current_provider)
+        if new_provider != None:
+            global_provider[name] = new_provider
```
Documentation for sure.

borkaehw commented on scala/private/phases/phase_final.bzl:
```diff
@@ -0,0 +1,38 @@
+
+def phase_test_final(ctx, p):
+    coverage_runfiles = p.coverage_runfiles.coverage_runfiles
+    coverage_runfiles.extend(p.write_executable)
```
Even though we could move it somewhere else, the return `struct` is still different. How do we use `phase_binary_final` here?

borkaehw commented on scala/private/phases/phase_final.bzl:
```diff
@@ -0,0 +1,38 @@
+        transitive_rjars = p.compile.rjars,  #calling rules need this for the classpath in the launcher
+    )
+
+def phase_library_final(ctx, p):
```
Providing actually code will be more clear what you mean about `phase_providers` and `rule_providers`, but I totally support the idea of using providers.

borkaehw commented on scala/private/phases/phase_compile.bzl:
```diff
@@ -0,0 +1,144 @@
+        ctx,
+        p,
+        _args.srcjars if hasattr(_args, "srcjars") else depset(),
+        _args.buildijar if hasattr(_args, "buildijar") else False,
```
Good suggestion.

Fixed.

borkaehw replied:
Fixed.

borkaehw commented on scala/private/phases/phase_compile.bzl:
```diff
@@ -0,0 +1,144 @@
+        p,
+        srcjars,
+        buildijar,
+        implicit_junit_deps_needed_for_java_compilation,
```
Added.

borkaehw commented on scala/private/phases/phase_compile.bzl:
```diff
@@ -0,0 +1,144 @@
+        deps_providers,
+    )
+
+    return struct(
```
Added.

borkaehw commented:
@ittaiz I have addressed all the comments and pushed some commits. Let me know if I missed something. I appreciate if you can "resolve conversation" if your question has been answered.

I am happy we have progress on discussion :smiley:

ittaiz commented:
@borkaehw re your general question- yes we're moving forward.
I'd like to get this in.
There are questions about the level of customization, maybe styling and some other small issues but in general I think phases is an improvement to what we have (I'm also still processing the extension mechanism which is interesting as well).

One big concern I still have is whether this will be very far away from starlark idioms 6 months down the line. 
I'd like to loop in some bazel google people but I'm concerned the size of the PR is too big. Maybe you can curate a few select links that demonstrate this idiom (something like lines 10-25 from a.bzl and 12-20 from b.bzl) and then I'll try to ping them here?
 
Finalizing the iterations will take some more time but I'm trying to put whatever time I have into it (including a bit on the weekends).

ittaiz commented:
re stale- I was just in the middle and so didn't update the files since I didn't want a context switch. I'll review everything again.

borkaehw commented:
Here are the files that could demonstrate how phases work.
1) [scala/private/phases/api.bzl](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-c7801487dc5533d588f77b1828ac279e) : the API for phases mechanism. `run_phases` is the core of rule implementation. `extras_phases` and `_adjust_phases` are how we add new phases via rule macros.
2) [scala/private/rules/scala_binary.bzl](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-1d6a67bea1f64739a071906e3682455b) : an example how we use phases in a rule implementation. It shows the cleanness of the code becomes and the easiness to swap/insert phases. `make_scala_binary` is the public API to customize phases.
3) [test/phase/customizability_test.bzl](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-e0c5e662d1284b5be0e7286a02f03448) and [test/phase/phase_customizability_test.bzl](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-1b5479820ab0b537c4c215131aa495f0) : these are examples how to extend phases. Also, this [commit](https://github.com/bazelbuild/rules_scala/pull/874/commits/1064b4c67460eae8847ab77c55cd04770bf376cc) from https://github.com/bazelbuild/rules_scala/pull/874 is also a good example of phase extension.

borkaehw commented:
Thanks for doing another round of review. I will work on documentation and tests once we have finalized everything. :smiley:

ittaiz commented:
Still reviewing. I have to say that I just saw one significant advantage of the declarative nature of the phases!
I was reviewing the specific commits you added (split init and so on) and then noticed that the scala binary rules don't collect exports jars or srcjars. I have a hunch the srcjars issue is a bug and the exports might be a missing feature. 
We should talk about that but the bigger thing is that the phases made it easy to see it. Awesome!

One question this triggered though- how should I go about aggregating a few phases? I suddenly thought a drawback of having fine grain phases is that there is duplication between the rules. Often it's ok because it's high level enough but sometimes I think you'd like to expose an aggregate (sort of like init but in case you have a clear name and logical grouping).
I thought one way to do this is extract a method that returns an array of phases and so the rules could call that method + other explicit phases.
WDYT?

borkaehw commented:
I think an array of phases will help. However, it introduces new question: how do we customize a method/aggregate? If it's a pre-defined aggregate, we lose the strength of phases but only avoid a small chunk of duplication. If users can customize the aggregate, we will have even more complicated API to customize individual phase and aggregate. Even though some phases are shared across all the rules, they still call a slightly different methods (like `phase_binary_compile` and `phase_library_compile`).

I can see the benefit of being able to group phases, but I feel like we sacrifice more or bring more overhead to achieve it. Currently not sure how it will work out.

ittaiz commented:
The method is an implementation detail.
A way to save internal code duplication.
Because it returns an array of phases the input to runPhases is exactly the
same and so the user can still customize it exactly the same.
What am I missing?
The only change is that for us if we need to add a new phase in that
“dependency collection aggregate example” we don’t need to add it to all
places.

On Wed, 20 Nov 2019 at 21:43 Bor Kae Hwang <notifications@github.com> wrote:

> I think an array of phases will help. However, it introduces new question:
> how do we customize a method/aggregate? If it's a pre-defined aggregate, we
> lose the strength of phases but only avoid a small chunk of duplication. If
> users can customize the aggregate, we will have even more complicated API
> to customize individual phase and aggregate. Even though some phases are
> shared across all the rules, they still call a slightly different methods
> (like phase_binary_compile and phase_library_compile).
>
> I can see the benefit of being able to group phases, but I feel like we
> sacrifice more or bring more overhead to achieve it. Currently not sure how
> it will work out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/865?email_source=notifications&email_token=AAKQQFYCYS6GQ2MSR74G4QLQUWHMZA5CNFSM4JEL2CEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEUHLXI#issuecomment-556299741>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKQQF2T5YQVRIBL3C7UMLTQUWHMZANCNFSM4JEL2CEA>
> .
>
-- 


*Ittai Zeidman*



Cell: 054-6735021

40 Hanamal street, Tel Aviv, Israel


<http://www.wix.com>

borkaehw commented:
I might have more clear idea about what you are saying but still don't have the whole picture. Would you mind to create a PR to this PR with your proposed method? It's easier to see the code.

SrodriguezO commented:
@ittaiz's  suggestion is to minimize code duplication. For example, if you have
```
rule A phases: [a, b, c, d]
rule B phases: [a, b, c, e]
rule C phases: [a, b, f, c]
```
you can simplify this by 
```
abc_group = [a, b, c] (or a method that returns this array)
rule A phases: abc_group + d
rule B phases: abc_group + e
rule C phases: abc_group + f (unless f occurring before c mattered)
```

Am I understanding this correctly? If so, I do think this would be beneficial because it'd make things easier to maintain (as Ittai mentioned, changes would be more centralized). It'd also make the rule implementations more readable (if we're careful about grouping phases in a logical manner). I don't think we'd make aggregates easily extensible by consumers since they might shoot themselves in the foot, but that's something we can flesh out more later.

While I do think this is valuable, I think it should be done as a separate PR after the main refactor to avoid scope creep (especially since there's a bit of fleshing out to do).

ittaiz commented:
Thanks Sergio for this clarification!
1. This grouping is completely an implementation detail of rules_scala and
in no means an API.
2. This grouping needs indeed to be very careful and semantic. We want
abstractions. Indirection is worse than duplication.
3. We definitely don’t want to add anymore to this PR. I do want however to
understand if we’re aligned about the architecture and future options. Even
my comment about return type of providers and such which is crucial and
must happen should probably also happen in an additional PR.

On Fri, 22 Nov 2019 at 19:41 Sergio Rodriguez Orellana <
notifications@github.com> wrote:

> @ittaiz <https://github.com/ittaiz>'s suggestion is to minimize code
> duplication. For example, if you have
>
> rule A phases: [a, b, c, d]
> rule B phases: [a, b, c, e]
> rule C phases: [a, b, f, c]
>
> you can simplify this by
>
> abc_group = [a, b, c] (or a method that returns this array)
> rule A phases: abc_group + d
> rule B phases: abc_group + e
> rule C phases: abc_group + f (unless f occurring before c mattered)
>
> Am I understanding this correctly? If so, I do think this would be
> beneficial because it'd make things easier to maintain (as Ittai mentioned,
> changes would be more centralized). It'd also make the rule implementations
> more readable (if we're careful about grouping phases in a logical manner).
> I don't think we'd make aggregates easily extensible by consumers since
> they might shoot themselves in the foot, but that's something we can flesh
> out more later.
>
> While I do think this is valuable, I think it should be done as a separate
> PR after the main refactor to avoid scope creep (especially since there's a
> bit of fleshing out to do).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/865?email_source=notifications&email_token=AAKQQF5Z4DROQUEK4GCFHQDQVAKTXA5CNFSM4JEL2CEKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE6LA2I#issuecomment-557625449>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKQQF64RAOUYJBIIKRYF3LQVAKTXANCNFSM4JEL2CEA>
> .
>
-- 


*Ittai Zeidman*



Cell: 054-6735021

40 Hanamal street, Tel Aviv, Israel


<http://www.wix.com>

borkaehw commented:
In this case, I totally agree with the approach. We could talk about the details in future PRs.

ittaiz commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,51 @@
+def run_phases(ctx, builtin_customizable_phases, fixed_phase):
+    phase_providers = [
+        p[_ScalaRulePhase]
+        for p in ctx.attr._phase_providers
```
many places call provider as p, should it be provider to disabmiguate from phase?

borkaehw replied:
I suppose you meant the other way around "should it be phase to disabmiguate from provider?". I also worry I might forget what all this `p`s mean later, I decided to use more semantic name in this file.

borkaehw commented on scala/private/rules/scala_library.bzl:
```diff
@@ -195,14 +133,23 @@ def scala_library_suite(
+        [
+            ("scalac_provider", phase_scalac_provider),
+            ("collect_srcjars", phase_collect_srcjars),
+            ("collect_exports_jars", phase_collect_exports_jars),
```
I see what you are saying. I did it intentionally since the input seems to not depend on intermediate result. I suppose I can group it into `phase_init`. As we are breaking up `phase_init`, there is no good reason to change the order. I will put it back.

borkaehw replied:
Fixed.

borkaehw commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,51 @@
+    ]
+
+    if phase_providers != []:
+        builtin_customizable_phases = _adjust_phases(builtin_customizable_phases, [p for pp in phase_providers for p in pp.phases])
```
It makes sense. Fixed.

borkaehw commented on scala/private/phases/phase_compile.bzl:
```diff
@@ -0,0 +1,149 @@
+    )
+    return phase_common_compile(ctx, p, args)
+
+def phase_common_compile(ctx, p, _args = struct()):
```
It has been made private.

borkaehw commented:
> Double customizations

Ones with `_args` have been made private. `phase_common_*` will be the public interface for default arguments, `_phase_default_*` will be the private function with `_args` for the actual functionality.

> Grouping of different rule phases into single file

Glad we have agreement on this.

> Re-export of symbols

`scala.bzl` and `providers.bzl` have been moved (to new files) to `scala/advanced_usage/`. The files in `scala/private/phases/` remain in the same place. I suppose even they are power users, they will never need to access interfaces in `scala/private/phases/`. See [test/phase/customizability_test.bzl](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-e0c5e662d1284b5be0e7286a02f03448) for example. Let me know if you have other thought.

> Return type of phase

I totally agree on this. :+1:

It looks like we are getting closer to merging PR. I would like to start writing documentations, both in-line comments and doc files. I see we have a `README.md` and a bunch of `.md`s in `docs/`, I appreciate if you can point me to where should we put documents of customizable phases.

We are also upgrading CI to Bazel 1.1.0 and 1.2.0 (soon). Do you want me to rebase for compatibility check?

long-stripe commented:
MERGEBOT_SUBMIT

andyscott commented:
For whatever it's worth, I also approve of these changes. But I'm strongly biased 😂.

borkaehw commented:
Added some in-line comments. I will appreciate some helps for the documentation for individual phase, I am not 100% sure what each step is doing.

New doc file `docs/customizable_phase.md` is added. I have not finished it yet but suggestions for the outline and what contents we should include would be nice. I will push more commits as I come out with more contents.

Force push to keep this PR up-to-date, commits are kept.

borkaehw commented:
@ittaiz @long-stripe I have added the documentation. Please take a look.
[README](https://github.com/lucidsoftware/rules_scala_bazelbuild/tree/customizable-phase#advanced-configurable-rules)
[Customizable phase doc](https://github.com/lucidsoftware/rules_scala_bazelbuild/blob/customizable-phase/docs/customizable_phase.md)

I will work on writing more tests.

borkaehw commented:
@ittaiz I have finished the work for documentation and tests. Please take a look and let me know what would be the next step toward merging. Thanks.

ittaiz commented:
Small update (but valuable)- I ran this branch across Wix's codebase and everything is green!
I still have work here :( but it's a good sign
@ianoc @andyscott it would be very beneficial to get the same feedback from Stripe's repos

ittaiz commented:
@kevingessner @softprops and others are welcome to test this out and give feedback before we merge it in

ittaiz commented:
btw, I love the fact that after we moved the phases and such to advanced usage then it's very clear to see that the API hasn't changed. Only changes to `advanced-usage`, `private/phases` and tests of that. Nice.

long-stripe commented:
I just tried out https://github.com/bazelbuild/rules_scala/pull/865/commits/5b6f4ae0d6ae23faf3daa662ac47fdc73e7eaf4e with the default configuration on a branch against our CI and got ✅.

I'll have limited bandwidth coming up over the holidays, but perhaps @andyscott can check it out further with respect to phase customization?

borkaehw commented on docs/customizable_phase.md:
```diff
@@ -0,0 +1,148 @@
+## Who needs customizable phase
+Customizable phase is an advanced feature for people who want the rules to do more. If you are an experienced Bazel rules developer, we make this powerful API public for you to do custom work without impacting other consumers. If you have no experience on writing Bazel rules, we are happy to help but be aware it may be frucstrating at first.
+
+If you don't need to customize your rules and just need the default setup to work correctly, then just load the following file for default rules:
```
Good suggestion. Fixed.

borkaehw commented on scala/private/phases/api.bzl:
```diff
@@ -0,0 +1,86 @@
+        return phases
+    phases = phases[:]
+
+    # relation: the position to add a new phase
```
Not sure how much we want to include in `providers.bzl`, we have details [here](https://github.com/bazelbuild/rules_scala/pull/865/files#diff-3152e63cc3552788493d47c5dde927a2R67). But I still added some.

borkaehw commented:
Thank you all for your effort. Happy to see it moving forward smoothly. I will conclude the follow-up tasks in a document.

borkaehw commented:
@ittaiz I have summarized the follow-up actions in this doc. Please take a look and let me know if I missed something.
https://docs.google.com/document/d/1_oWoZ9T4ya82M3NU6Liew5x2IGoZMzZoEkFDknMKix4/edit?usp=sharing

Last thing, since we borrowed the phase architecture from [higherkindness/rules_scala](https://github.com/higherkindness/rules_scala), @andyscott is the original author, we have to figure out the attribution to give credit to Andy. Please hold off on merging before we hear back from him. I have contacted him, but it's holiday season. I will also be on vacation the next few weeks, I won't be completely offline but the response may take longer. @SrodriguezO is another you can contact with. Thanks!

andyscott commented:
I wouldn't mind being officially listed in AUTHORS, as this PR pulls over the fundamental design components of the alternative rules. I worked extensively on this design in my own free time.

That being said, feel free to merge this PR when ready.

andyscott commented:
I went ahead and opened a PR ;)

https://github.com/bazelbuild/rules_scala/pull/903

andyscott commented:
Can we merge this soon and then iterate in follow up PRs?

ittaiz commented:
@andyscott thats the plan.
I’m merging it this Wednesday.

ittaiz commented:
@borkaehw @SrodriguezO out of interest- did lucid ever move to bazelbuild/rules_scala or remain on higherkindness?
